src/main/java/io/micronaut/build/MicronautQualityChecksParticipantPlugin.java

```diff
@@ -1,20 +1,70 @@
 package io.micronaut.build;
 
+import static io.micronaut.build.MicronautQualityChecksAggregatorPlugin.COVERAGE_REPORT_TASK_NAME;
+
+import java.io.File;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
+import org.gradle.api.plugins.quality.CheckstyleExtension;
+import org.gradle.api.plugins.quality.CheckstylePlugin;
 import org.gradle.testing.jacoco.plugins.JacocoPlugin;
+import org.sonarqube.gradle.SonarQubeExtension;
 
-@SuppressWarnings("UnstableApiUsage")
 public class MicronautQualityChecksParticipantPlugin implements Plugin<Project> {
 
     @Override
     public void apply(final Project project) {
+        final MicronautBuildExtension micronautBuild = project.getExtensions().findByType(MicronautBuildExtension.class);
+
+        configureCheckstyle(project, micronautBuild);
+        configureSonar(project);
         configureJacoco(project);
     }
 
-    private void configureJacoco(final Project project) {
-        project.getPluginManager().withPlugin("java", p -> {
-            project.getPluginManager().apply(JacocoPlugin.class);
+    private void configureCheckstyle(final Project project, final MicronautBuildExtension micronautBuildExtension) {
+        project.afterEvaluate(p -> {
+            p.getPluginManager().apply(CheckstylePlugin.class);
+            final CheckstyleExtension checkstyle = p.getExtensions().findByType(CheckstyleExtension.class);
+            if (checkstyle != null) {
+                checkstyle.setConfigFile(p.file("${rootDir}/config/checkstyle/checkstyle.xml"));
+                checkstyle.setToolVersion(micronautBuildExtension.getCheckstyleVersion().get());
+
+                // Per submodule
+                checkstyle.setMaxErrors(1);
+                checkstyle.setMaxWarnings(10);
+
+                checkstyle.setShowViolations(true);
+            }
+
+            p.getTasks().named("checkstyleTest").configure(task -> task.setEnabled(false));
+            p.getTasks().named("checkstyleMain").configure(task -> {
+                task.dependsOn("spotlessCheck");
+                project.getRootProject().getTasks().named(COVERAGE_REPORT_TASK_NAME).configure(t -> t.dependsOn(task));
+            });
+
         });
     }
+
+    private void configureSonar(final Project project) {
+        if (System.getenv("SONAR_TOKEN") != null) {
+            project.getRootProject().getPluginManager().withPlugin("org.sonarqube", p -> {
```

Comment:
This whole block looks fishy. For every participant, this mutates the root project configuration, which is probably going to fail. In addition, this overwrites the `sonar.java.checkstyle.reportPaths` to the current project checkstyle report path, meaning that other projects are going to be ignored.

I think it should be wired the other way around: if there's no aggregate checkstyle report, we should either generate one, or do the same as we do for javadocs/jacoco, which is to say expose checkstyle results and consume them from the root project (that is, the root project consumes checkstyle reports from dependencies, the subprojects. Not the subproject pushing their configurations to the root). Also it's wrong because the configuration will _only_ work if the checkstyle report file is present, that is if the task was called before.

Reply:
In which way this block mutates the root project configuration? My intention (and please correct me if the code is doing something else) is to set a property in a subproject's extension. Each one gets its own value. Note that this is working (as in producing the expected result) correctly.

Also, note that Checkstyle report "aggregation" doesn't work in the root project. It is the first thing I tried, since `sonar.java.checkstyle.reportPaths` can actually be a comma-separated list of reports. However, the Sonar plugin doesn't work well with that, since the root project doesn't have any source associated, so it can't process any of the files contained in the report.

Regarding your last sentence, I did set a dependency on the checkstyle task, but mistakenly associated to the coverage report task. Fixed in the last commit. Still, for projects that lack a Checkstyle report (because it's missing, disabled, or whatever), the `sonar.java.checkstyle.reportPaths` is not defined, and Checkstyle won't attempt to parse that report.

Reply:
Oh I see, I got confused by the `project.getRootProject().getPluginManager().withPlugin("org.sonarqube", p -> {` block.

Reply:
We can't do the _configuration_ of the task if the checkstyle file exists. The checkstyle file will be defined once the task is executed. So we should always configure it to use that XML report file, whether it exists or not. As for the dependency, there should actually be a dependency between the root project "sonar" task and considered subprojects' `checkstyle` task, which is clearly something that the sonar plugin must be doing. If it does, it's probably doing it wrong. If it doesn't, then we can workaround this.